File: Day #442/solution.py
# Day 442: Cartesian tree (min-heap ordered, in-order == S) built with a
# monotonic stack in O(n) time, O(n) space.


class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def build_cartesian(s):
    stack = []
    for v in s:
        node = Node(v)
        last = None
        while stack and stack[-1].val > v:
            last = stack.pop()
        node.left = last
        if stack:
            stack[-1].right = node
        stack.append(node)
    return stack[0] if stack else None


def show(n, prefix="", tag=""):
    if n is None:
        return
    print(prefix + tag + str(n.val))
    show(n.left, prefix + "  ", "L-- ")
    show(n.right, prefix + "  ", "R-- ")


if __name__ == "__main__":
    root = build_cartesian([3, 2, 6, 1, 9])
    show(root)
    # 1
    #   L-- 2
    #     L-- 3
    #     R-- 6
    #   R-- 9

File: Day #442/solution.go
// Day 442: Cartesian tree (min-heap ordered, in-order == S) built with a
// monotonic stack in O(n) time, O(n) space.
package main

import "fmt"

type Node struct {
	val         int
	left, right *Node
}

func buildCartesian(s []int) *Node {
	var stack []*Node
	for _, v := range s {
		node := &Node{val: v}
		var last *Node
		for len(stack) > 0 && stack[len(stack)-1].val > v {
			last = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
		}
		node.left = last
		if len(stack) > 0 {
			stack[len(stack)-1].right = node
		}
		stack = append(stack, node)
	}
	if len(stack) == 0 {
		return nil
	}
	return stack[0]
}

func show(n *Node, prefix, tag string) {
	if n == nil {
		return
	}
	fmt.Println(prefix + tag + fmt.Sprint(n.val))
	show(n.left, prefix+"  ", "L-- ")
	show(n.right, prefix+"  ", "R-- ")
}

func main() {
	root := buildCartesian([]int{3, 2, 6, 1, 9})
	show(root, "", "")
	// 1
	//   L-- 2
	//     L-- 3
	//     R-- 6
	//   R-- 9
}
